### PR TITLE
feat: loading screen

### PR DESCRIPTION
**Pull request by LPLafontaineB**
### Description (*)

This PR adds a loading screen that appears when changing scenes.
### Issue Number(s) (*)

Fixes issue(s): [MTT-1960](https://jira.unity3d.com/browse/MTT-1960) [MTT-1746](https://jira.unity3d.com/browse/MTT-1746)
### Manual testing scenarios

1. Start a game with a host
2. Join with a client
3. Move the client around
4. Disconnect the client
5. Reconnect the client

The host and the client should see the loading screen when starting the game, then when changing scenes to the CharSelect scene and the BossRoom scene. Those loading screens should hide the slight delay before player characters are spawned, during which the additive scene is not loaded in the BossRoom scene, the camera is at its default position, and the UI is at its default state. This should be more evident during step 5, which should hide the camera movement described in [MTT-1746](https://jira.unity3d.com/browse/MTT-1746)

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by SamuelBellomo** [comment]
mmm the diff contains changes from additive scene loading :( I tried forcing github to do a reload of the diff (by switching base branch), but that didn't fix it. @LPLafontaineB you might want to look into this before we review

**Comment by LPLafontaineB** [comment]
> mmm the diff contains changes from additive scene loading :( I tried forcing github to do a reload of the diff (by switching base branch), but that didn't fix it. @LPLafontaineB you might want to look into this before we review

Yeah, it probably happened because I was working on this in parallel and started from the additive scene branch instead of develop. I'll see how I can fix this

**Comment by LPLafontaineB** [comment]
Should be OK now @SamuelBellomo

**Comment by SamuelBellomo** [comment]
just tested real quick, I still see unloaded maps after the spinning load screen.

**Comment by LPLafontaineB** [comment]
> just tested real quick, I still see unloaded maps after the spinning load screen host side.

I expected that the fade-out could hide that since I didn't notice it during my tests. In that case, to do it properly I might have to create a more complex solution since I don't think we want loading screens each time a new additive scene is loaded after the actual game starts. I'll return this PR to draft while I work on that

File: Assets/BossRoom/Scripts/Client/UI/PostGameUI.cs
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Client/UI/PostGameUI.cs:80]: shouldn't the switch to CharSelect remain synced with the other clients and the host?

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs
Comment (LPLafontaineB) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs:14]: That's what I did originally, but it doesn't work as well when loading the BossRoom scene. In that case, it would stop the loading screen before the first additive scene loads, and before the player characters spawn, which is what it was supposed to hide. Basically, I need a way to prevent the loading screen from stopping when the BossRoom scene is loaded and have it wait for all additive scenes. But I agree it is not an elegant solution.

I could get the RTT and add it as a delay before stopping the loading screen, which should give enough time to start loading the additive scenes, but I'm not sure how to do this cleanly yet. Plus I'm not a fan of adding these kinds of delays, it feels like a workaround. I will try to think of something else.

Comment (LPLafontaineB) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs:14]: Nevermind, it works. The issue I remembered was probably due to me using the wrong events in the beginning of this PR.

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs:11]:
```suggestion
        /// Adds loading screens by wrapping scene management APIs. It loads scene using the SceneManager, or, on
```
Just to put the emphasize on the right thing. Users with no loading screens don't need this class, correct?

File: Assets/BossRoom/Scripts/Client/Game/State/ClientCharSelectState.cs
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Client/Game/State/ClientCharSelectState.cs:406]: instead of adding single everywhere, couldn't we add a default value to SceneLoader's LoadScene?

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/GameNetPortal.cs
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/GameNetPortal.cs:251]: I'm curious why this was moved?

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs:8]: Non blocker nitpicking: That name tickles me. When I first saw it used in game code, I was wondering "why isn't Netcode's scene manager used anymore?"
and then I finally understood SceneLoader is a convenience wrapper that adds loading screen management to normal scene loading calls.
Maybe SceneLoaderWrapper? LoadingScreenManager?
I don't know, naming things is hard :/

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs:31]: I know Netcode used to have issues with DontDestroyOnLoad and scene switching. That's solved right? If you finish a game and try to start a new one, this should work fine correct?

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs:49]: public methods should have a quick doc.

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs:39]: does this need to be public?

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs
Comment (LPLafontaineB) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs:8]: It is hard :( I probably spent more time trying to name this class than coding this entire PR. It has this name because my original intention was to create a generic wrapper for loading scenes because there is no callback I could attach to to start the loading screen when offline. But in the end, the only use it has is to tell the loading screen when to start and when to stop.

I think SceneLoaderWrapper would make more sense. LoadingScreenManager would create confusion with those wrapper methods I think, since from the outside they still load scenes first.

File: Assets/BossRoom/Scripts/Client/Game/State/ClientCharSelectState.cs
Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Client/Game/State/ClientCharSelectState.cs:406]: We could, but I like having it that way, like how it is with NetworkSceneManager. It makes it more deliberate to chose to load a scene in Single mode.

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs
Comment (LPLafontaineB) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs:31]: It does not work fine. If I start a game, then quit, then start a new one, I get an exception on the client that says: `Unhandled RPC Exception: System.InvalidCastException: Specified cast is not valid.` when joining. It comes from `StopLoadingScreenClientRpc`.

I'll try to debug this

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs
Comment (LPLafontaineB) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs:39]: The event function it overrides in NetworkBehavior is public. The IDE gives me an error if I make it private, and if I just remove the override keyword I get a warning because it hides NetworkBehavior's method

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/GameNetPortal.cs
Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/GameNetPortal.cs:251]: It was because it would execute before AddOnSceneEventCallback could be called, but looking back at it I did not need to move it, just to put AddOnSceneEventCallback in the same place instead. I'll move it back, I think it is better how it was.

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs
Comment (SamuelBellomo) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs]: I'm open to other names too, SceneLoaderWrapper isn't great either

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs
Comment (SamuelBellomo) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs]: oh I forgot NetworkBehaviour implemented this. NVM nothing to see here

File: Assets/BossRoom/Scripts/Client/Game/State/ClientCharSelectState.cs
Comment (SamuelBellomo) [reply Assets/BossRoom/Scripts/Client/Game/State/ClientCharSelectState.cs]: And the static LoadScene method has default parameters, just like Unity's SceneManager does too https://docs.unity3d.com/ScriptReference/SceneManagement.SceneManager.LoadScene.html
This way, users that want simple calls can just use the default behaviour, without preventing them to be explicit about it (those that want to be explicit can still "override" the default param with the same default param)

Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Client/Game/State/ClientCharSelectState.cs]: I don't see any static LoadScene method with default parameters in NetworkSceneManager, only in Unity's SceneManager, am I looking in the wrong place? Either way, I'll add it some to this new method then

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs
Comment (LPLafontaineB) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs]: Pre.4 fixes this issue, but there is another one still. In short, if the host is shut down, then restarted, the NetworkObject is not re-spawned. I created a [github issue](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/issues/1574) for it. I think we should put this PR on hold in the meantime.

Comment (SamuelBellomo) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs]: When you say restarted and the NO is not respawned, you mean host side right? Cause I wouldn't expect a client to still be connected after a shutdown.
So when starting a 2nd game in the same process, the host can't have that object present a second time? yeah that's bad :(

Comment (LPLafontaineB) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs]: It actually happens on both the host and the client. After leaving a game, if any of the two game processes try to restart a new game, the NetworkObject will not be spawned. As you intuited, it doesn't make the game unplayable, it simply doesn't activate the loading screen on the client (it does still activate them on the host, though). However, if we shutdown the session a second time, we will be stuck in the main menu with an infinite loading screen. In short, the game is playable once normally, then a second time with no loading screen on the client, but not a third time.

If, however, we have a host or client that has started hosting or joined a game, then quits, then tries to join a game that is hosted by a new process, it will give use exceptions and fail to spawn the NetworkObject, which makes the game unplayable. The reverse case (a host that has a previous session joined by a new client) does not cause this issue.

Comment (SamuelBellomo) [reply Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoader.cs]: yeah ok, I agree, let's keep this on hold then

**Comment by LPLafontaineB** [comment]
On hold, waiting for [this issue](https://github.com/Unity-Technologies/com.unity.netcode.gameobjects/issues/1571) to be solved on the sdk

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/ClientGameNetPortal.cs
Comment (SamuelBellomo) [inline Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/ClientGameNetPortal.cs:238]: 1. ClientGameNetPortal shouldn't have to know about this, it's not its responsibility and just adds coupling.
2. method name is a bit weird, add...something usually means you're adding something and would expect a parameter. "Init" something might be better.

Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/ClientGameNetPortal.cs:238]: The issue with this is that to trigger the loading screen at the beginning of client syncrhonization, we need to register those callbacks before they happen. However, the NetworkSceneManager is created when StartClient / StartHost / StartServer is called, so we have to register them after this. OnClientConnected and OnNetworkSpawn happen after the synchronization has finished so it must be done before those callbacks. We currently don't have access to callbacks when starting clients, which is why I had to do it this way.

Comment (SamuelBellomo) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/ClientGameNetPortal.cs:238]: I'd create our own callback for this then, instead of having our connection management know about scene management initialization?

Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/ClientGameNetPortal.cs:238]: If we do that though, that means that we need to add coupling in the other direction, which means we will have to move the loading screen stuff out of the Utilities package. My initial goal was to create something generic enough to be able to be reused in other projects, but if it makes more sense that way, I can move it.

Comment (SamuelBellomo) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/ClientGameNetPortal.cs:238]: @pdeschain you have the arch cleanup initiative. What do you think of moving connection management to Utilities as well? To me it's the kind of stuff we'd want to reuse in future samples anyway

Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/ClientGameNetPortal.cs:238]: Should that be part of a future ticket instead? Or should we move all of that to the Shared assembly for now and evaluate later what should be moved to Utilities?

Comment (SamuelBellomo) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/ClientGameNetPortal.cs:239]: future ticket, just make sure it's tracked :)

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/ClientLoadingScreen.cs
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/ClientLoadingScreen.cs:87]: this should be a WaitForSeconds

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoaderWrapper.cs
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoaderWrapper.cs:134]: Why do we need to wait for the server to tell us to stop the loading screen? Do we have an event client side once we're done syncing? Why the back and forth?

**Comment by SamuelBellomo** [comment]
Removing the on hold label for review, however let's make sure we merge this only once we have pre.7 updated

File: Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/ClientGameNetPortal.cs
Comment (LPLafontaineB) [reply Assets/BossRoom/Scripts/Shared/Net/ConnectionManagement/ClientGameNetPortal.cs:239]: here it is: https://jira.unity3d.com/browse/MTT-3157

File: Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoaderWrapper.cs
Comment (SamuelBellomo) [inline Packages/com.unity.multiplayer.samples.coop/Utilities/SceneManagement/SceneLoaderWrapper.cs:133]:
```suggestion
                        // Send client RPC to make sure the client stops the loading screen after the server handles what it needs to after the client finished synchronizing, for example character spawning done server side should still be hidden by loading screen.
```

**Review by SamuelBellomo — APPROVED** [review]

**Review by fernando-cortez — DISMISSED** [review]

**Review by SamuelBellomo — DISMISSED** [review]

**Comment by LPLafontaineB** [comment]
Adding DoNotMerge label until BossRoom is updated to NGO pre.7

**Review by SamuelBellomo — APPROVED** [review]

**Review by fernando-cortez — APPROVED** [review]

**Pull request merged**
